File: src-tauri/src/agent/native_tools/knowledge_base/search_knowledge_base.rs
//! `search_knowledge_base` 工具（原生）— 语义检索指定知识库，返回上下文文本 + 结果 uiData。

use crate::agent::native_tools::common::{arg_i64, arg_str};
use crate::agent::native_tools::{NativeToolCtx, NativeToolOutcome};
use serde_json::{json, Value};

use super::common::{build_search_context, rag_service};

pub(crate) async fn search_knowledge_base_tool(
    _ctx: &NativeToolCtx<'_>,
    args: &Value,
) -> Result<NativeToolOutcome, String> {
    let query = arg_str(args, "query").unwrap_or_default();
    if query.trim().is_empty() {
        return Ok(NativeToolOutcome::Value {
            content: "Missing required parameter: \"query\". Please provide a search query.".to_string(),
            ui_data: None,
        });
    }
    let kb_id = arg_str(args, "knowledge_base_id").unwrap_or_default();
    if kb_id.is_empty() {
        return Ok(NativeToolOutcome::Value {
            content: "Missing required parameter: \"knowledge_base_id\". Use list_knowledge_bases to discover available knowledge bases and their IDs.".to_string(),
            ui_data: None,
        });
    }
    let top_k = arg_i64(args, "top_k").unwrap_or(5).clamp(1, 20) as usize;

    let service = rag_service()?;
    let kb_id_c = kb_id.clone();
    let query_c = query.trim().to_string();
    let results = tokio::task::spawn_blocking(move || service.query(&kb_id_c, &query_c, top_k))
        .await
        .map_err(|e| format!("Task join error: {}", e))?
        .map_err(|e| format!("Error searching knowledge base: {}", e))?;

    if results.is_empty() {
        return Ok(NativeToolOutcome::Value {
            content: format!("No relevant information found in knowledge base \"{}\" for query: \"{}\".", kb_id, query),
            ui_data: Some(json!({ "length": 0, "query": query })),
        });
    }

    let context = build_search_context(&results, &query, &kb_id);
    let ui_results: Vec<Value> = results
        .iter()
        .map(|r| {
            json!({
                "id": r.id,
                "document_name": r.document_name,
                "document_id": r.document_id,
                "score": r.score,
                "snippet": r.content.chars().take(200).collect::<String>(),
            })
        })
        .collect();

    Ok(NativeToolOutcome::Value {
        content: context,
        ui_data: Some(json!({
            "length": results.len(),
            "results": ui_results,
            "query": query,
            "knowledge_base_id": kb_id,
        })),
    })
}

File: src/infrastructure/tools/knowledge-base/search-knowledge-base.ts
/**
 * search_knowledge_base — 语义搜索知识库
 *
 * AI 可以在回答问题时主动检索知识库中的相关文档内容，
 * 就像 web_search 搜索互联网一样。
 */
import { toolRegistry } from '@/domain/tools'
import type { ToolExecutor, ToolResult } from '@/domain/tools/types'
import { t } from '@/ui/i18n'
import { ragService } from '@/services/rag-service'
import { buildSearchContext } from './common'
import { sliceHead } from '@/utils/text'

toolRegistry.register(
    'search_knowledge_base',
    (async (args: Record<string, any>, _ctx: any): Promise<ToolResult> => {
    const query = args.query
    if (!query || typeof query !== 'string' || query.trim() === '') {
      return {
        content:
          'Missing required parameter: "query". Please provide a search query.',
      }
    }

    const kbId = args.knowledge_base_id
    if (!kbId || typeof kbId !== 'string') {
      return {
        content:
          'Missing required parameter: "knowledge_base_id". ' +
          'Use list_knowledge_bases to discover available knowledge bases and their IDs.',
      }
    }

    const topK = Math.min(args.top_k ?? 5, 20)

    try {
      const result = await ragService.query(kbId, query.trim(), topK)

      if (!result.results || result.results.length === 0) {
        return {
          content: `No relevant information found in knowledge base "${kbId}" for query: "${query}".`,
          uiData: { length: 0, query },
        }
      }

      // 构建包含 document_id 的上下文（Rust 返回的 context 不含 document_id）
      const contextWithDocIds = buildSearchContext(result.results, query, kbId)

      return {
        content: contextWithDocIds,
        uiData: {
          length: result.results.length,
          results: result.results.map((r) => ({
            id: r.id,
            document_name: r.document_name,
            document_id: r.document_id,
            score: r.score,
            snippet: sliceHead(r.content, 200),
          })),
          query,
          knowledge_base_id: kbId,
        },
      }
    } catch (error: any) {
      return {
        content: `Error searching knowledge base: ${error.message || String(error)}`,
      }
    }
  }) as ToolExecutor,
    t('搜索知识库'),
)
